src/Evolution/Systems/GeneralizedHarmonic/Equations.hpp

```diff
@@ -127,4 +130,135 @@ struct ComputeNormalDotFluxes {
 };
+
+/*!
+ * \ingroup NumericalFluxesGroup
+ * \brief Computes the generalized-harmonic upwind flux
+ *
+ * The upwind flux in general is given by Eq. (6.3) of \cite Teukolsky2015ega:
+ * \f{eqnarray}{
+ * F^* = S \Lambda^+ S^{-1} u^- + S \Lambda^{-1} S^{-1} u^+,
+ * \end{eqnarray} \f}
+ * where \f$u\f$ is a vector of the evolved variables, \f$S^{-1}\f$
+ * maps the evolved variables into the characteristic variables \f$S^{-1}u\f$,
+ * \f$S\f$ maps the characteristic variables into the evolved variables,
+ * and $\Lambda$ is a diagonal matrix of the characteristic speeds.
+ * Here, \f$S^{-1}u^-\f$ represents the characteristic variables at the
+ * boundary internal to the element, \f$S^{-1}u^+\f$ represents the
+ * characteristic variables at the boundary but in the external, neighboring
+ * element, \f$\Lambda^+\f$ is a diagonal matrix whose nonzero entries
+ * are the characteristic speeds that are positive ("outgoing", i.e.
+ * leaving the element), and \f$\Lambda^-\f$ is a diagonal matrix whose
+ * nonzero entries are the characteristic speeds that are negative
+ * ("incoming", i.e. entering the element).
+ *
+ * This function implements the upwind flux for the generalized harmonic
+ * system. Specifically, it computes \f$\Lambda^+\f$ and \f$\Lambda^-\f$
+ * from the characteristic speeds of the element. Then, it computes the
+ * combination \f$\Lambda^+ S^{-1} u^- + \Lambda^- S^{-1} u^+\f$. Finally, it
+ * applies \f$S\f$ by converting the result back from characteristic to
+ * evolved variables, using the unit normal vector of the element.
+ */
+template <size_t Dim>
+struct UpwindFlux {
+ public:
+  using options = tmpl::list<>;
+  static constexpr OptionString help = {
+      "Computes the generalized harmonic upwind flux. It requires no "
+      "options."};
+
+  // clang-tidy: non-const reference
+  void pup(PUP::er& /*p*/) noexcept {}  // NOLINT
+
+  // This is the data needed to compute the numerical flux.
+  // `dg::SendBoundaryFluxes` calls `package_data` to store these tags in a
+  // Variables. Local and remote values of this data are then combined in the
+  // `()` operator.
+
+  using package_tags = tmpl::list<
+      Tags::UPsi<Dim, Frame::Inertial>, Tags::UZero<Dim, Frame::Inertial>,
+      Tags::UPlus<Dim, Frame::Inertial>, Tags::UMinus<Dim, Frame::Inertial>,
+      ::Tags::CharSpeed<Tags::UPsi<Dim, Frame::Inertial>>,
+      ::Tags::CharSpeed<Tags::UZero<Dim, Frame::Inertial>>,
+      ::Tags::CharSpeed<Tags::UPlus<Dim, Frame::Inertial>>,
+      ::Tags::CharSpeed<Tags::UMinus<Dim, Frame::Inertial>>,
+      Tags::ConstraintGamma2, Tags::UnitNormalOneForm<Dim, Frame::Inertial>>;
+
+  // These tags on the interface of the element are passed to
+  // `package_data` to provide the data needed to compute the numerical fluxes.
+  using argument_tags = tmpl::list<
+      Tags::UPsi<Dim, Frame::Inertial>, Tags::UZero<Dim, Frame::Inertial>,
+      Tags::UPlus<Dim, Frame::Inertial>, Tags::UMinus<Dim, Frame::Inertial>,
```

Comment:
Something is confusing me here with the $u^+$ and $u^-$. These shouldn’t be argument/package tags, since they represent the field value on either side of the interface. The packaging is done on both sides separately, then communicated and combined in the `operator()`. So I think the tags should be more like the following:
- The argument tags should be $u$ and the char speeds
- The packaged tags should either be those again, or even better as far as you can compute with the data that’s available on that side of the interface alone (e.g. $S \Lambda S^-1 u$ if that’s possible)
- The `operator()` then takes the packaged data from both sides of the interface as arguments (as the comments in the code, presumably copied from the elliptic code, describe). So they will be $u^-$ and $u^+$ automatically. The current code with e.g. `u_minus_int` and `u_minus_ext` does not make sense to me, since „minus“ means „internal“ and „plus“ means „external“, you don’t need both.

This also means that I don’t understand why there is are UPlus and UMinus tags in GeneralizedHarmonic - in a DataBox there will only ever be data for that element, and only in the flux is data from two elements combined. If I‘m getting something wrong here, please correct me.

Reply:
Thanks for your comments, @nilsleiffischer! In Generalized harmonic, the char fields are named U^+, U^-, U^0, and U^\Psi. This use of the plus/minus superscript does not refer to evaluating on one side of the interface or the other.